x/mock/types.go

```diff
@@ -0,0 +1,58 @@
+package mock
+
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/crypto"
+)
+
+type (
+	// TestAndRunTx produces a random transaction, and check the state
+	// transition. It returns a descriptive message "action" about what this
+	// random tx actually did, for ease of debugging.
+	TestAndRunTx func(
+		t *testing.T, r *rand.Rand, app *App, ctx sdk.Context,
+		privKeys []crypto.PrivKey, log string,
+	) (action string, err sdk.Error)
+
+	// RandSetup performs the random setup the mock module needs.
+	RandSetup func(r *rand.Rand, privKeys []crypto.PrivKey)
+
+	// An Invariant is a function which tests a particular invariant.
+	// If the invariant has been broken, the function should halt the
+	// test and output the log.
+	Invariant func(t *testing.T, app *App, log string)
+)
```

Comment:
As far as I can tell these three types should also be packaged together a a single object which gets passed around as opposed to passing around three objects individually as is being done right now, can't think of a time when you'd want one but not the others?

Reply:
I think it may be better split apart for composability reasons. This will give more freedom for the different types of long-running tests being ran. Eventually I'd like the test writer to be able to give different weightings for the probability of certain types of txs occuring, and that wouldn't be do-able if each module just exported a single "TestThisModule" struct. 

This also lets the test writers use functions like `PeriodicInvariant` if they want a particular invariant check ran less often. (as it could be slowing down the test) 

I can also see a module having multiple message types, but I don't want to test all of them in my mock application.

I'm not sure if could be done cleanly if they were all packaged into a single struct. (Without resorting to exporting each individual test/invariant, at which point they could be seperate)

Reply:
Let's keep as is for this PR

Reply:
More thoughts in next comment